pkg/arvo/app/thread-watch-hook.hoon

```diff
@@ -0,0 +1,199 @@
+      %watch
+    ?:  ?|(=(app.observer %spider) =(app.observer %thread-watch-hook)) 
+      ~|('we avoid infinite loops' !!)
+    ?:  (~(has in vals) observer)
+      ~|('duplicate observer' !!)
+    :_  this(observers (~(put by observers) now.bowl observer))  
+    :_  ~
+    :*  %pass
+        /observer/(scot %da now.bowl)
+        %agent
+        [our.bowl app.observer]
+        %watch
+        path.observer
+    == 
+  ::
+      %ignore
+    ?.  (~(has in vals) observer)
+      ~|('cannot remove nonexistent observer' !!)
+    =/  key  (got-by-val observers observer)
+    :_  this(observers (~(del by observers) key))
+    :_  ~
+    :*  %pass
+        /observer/(scot %da key)
+        %agent
+        [our.bowl app.observer]
+        %leave
+        ~
+    == 
+  ==
+::
+++  on-agent
+  |=  [=wire =sign:agent:gall]
+  ^-  (quip card _this)
+  ?-    -.sign
+      %kick
+    ?+    wire  (on-agent:def wire sign)
+        [%thread-result @ ~]  [~ this]
+        [%observer @ ~]
+      =/  =time  (slav %da i.t.wire)
+      =/  =observer:sur  (~(got by observers) time)
+      :_  this
+      :_  ~
+      :*  %pass
+          wire
+          %agent
+          [our.bowl app.observer]
+          %watch
+          path.observer
+      == 
+    ==
+  ::
+      %poke-ack
+    ?.  ?=([%thread-start @ @ ~] wire)
+      (on-agent:def wire sign)
+    ?~  p.sign  [~ this]
+    =/  =time  (slav %da i.t.wire)
+    =/  tid  (slav %uv i.t.t.wire)
+    =/  =observer:sur  (~(got by observers) time)
+    :_  this(observers (~(del by observers) time))
+    :~  :*  %pass
+            [%observer i.t.wire ~]
+            %agent
+            [our.bowl app.observer]
+            %leave
+            ~
+        ==
+        :*  %pass
+            wire
+            %agent
+            [our.bowl app.observer]
+            %leave
+            ~
+        == 
+    ==
+  ::
+      %watch-ack
+    ?~  p.sign  [~ this]
+    ?+    wire  (on-agent:def wire sign)
+        [%thread-result @ ~]  [~ this]
+        [%observer @ ~]
+      =/  =time  (slav %da i.t.wire)
+      [~ this(observers (~(del by observers) time))]
+    ==
+  ::
+      %fact
+    ?+    wire  (on-agent:def wire sign)
+        [%thread-result @ ~]
+      ?.  =(p.cage.sign %thread-fail)
+        [~ this]
+      =/  =time  (slav %da i.t.wire)
+      =/  =observer:sur  (~(got by observers) time)
+      :_  this(observers (~(del by observers) time))
+      :_  ~
+      :*  %pass
+          [%observer i.t.wire ~]
+          %agent
+          [our.bowl app.observer]
+          %leave
+          ~
+      == 
+    ::
+        [%observer @ ~]
+      =/  =time  (slav %da i.t.wire)
+      =/  =observer:sur  (~(got by observers) time)
+      =/  tid  (scot %uv (sham eny.bowl))
+      :_  this
+      :~  :*  %pass
+              [%thread-start i.t.wire tid ~]
+              %agent
+              [our.bowl %spider]
+              %poke
+              %spider-start
+              !>([~ `tid thread.observer (slop q.cage.sign !>(~))])
```

Comment:
Maybe it's the right time to change this behaviour

Reply:
Not yet :)